Add ensure capacity for enumerable types.

TimothyMakkison opened the pull request:
Complete #251 for enumerable types. I'm not sure how Explicit `Count` values should be treated, right now it gets the runtime type using the `is` operator. I think it should either try not to get the value or just cast the value so the explicit property can be accessed.

- Added `EnsureCapacityBuilder`, `CanEnsureCapacity` checks if ensure capacity can be used, storing the relevant information in `EnsureCapacityInfo` to be used by `TryBuildEnsureCapacityStatement` when the syntax is built.
  - If both source and target have compile time sizes `TryBuildEnsureCapacityStatement` generates a simple call to `EnsureCapacity`.
  - If source does not have a length/count property, `is` statements are created to check the runtime type, invoking `EnsureCapacity` with the resulting type.
  - Not sure where `EnsureCapacityBuilder` should be placed.
- Added tests, waiting for integration tests from github actions.

latonz commented on src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs:
```diff
@@ -0,0 +1,131 @@
+
+namespace Riok.Mapperly.Descriptors.Mappings.ExistingTarget;
+
+public class EnsureCapacityInfo
```
This code base usually follows one type per file approach.

To match existing architecture convert these properties to private fields and create an exposed method `BuildEnsureCapacity()` which creates the syntax (content of `TryBuildEnsureCapacityStatement`) (similar to how the object factories and mappings work).

latonz commented on src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs:
```diff
@@ -0,0 +1,131 @@
+        }
+        if (value.HasImplicitInterfaceProperty(types.ICollectionT, "Count") || value.HasImplicitInterfaceProperty(types.IReadOnlyCollectionT, "Count"))
+        {
+            expression = "Count";
```
Introduce const's and probably nameofs for these strings.

latonz commented on src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs:
```diff
@@ -0,0 +1,131 @@
+        // this is used if source is IEnumerable
+        var targetCount = MemberAccess(target, state.TargetAccessor);
+
+        var ifCollection = IfIsTypeEnsureCapacityStatement("collection", state.CollectionType!, ctx, target, targetCount);
```
If you use named identifiers in the generated code always make sure to use `ctx.NameBuilder.New` to ensure that the name is unique.

latonz commented on src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs:
```diff
@@ -0,0 +1,131 @@
+            return true;
+        }
+
+        // generate check for runtime type, calling EnsureCapacity if it is a collection/readonlyCollection
+        // this is used if source is IEnumerable
```
If the target framework supports `Enumerable.TryGetNonEnumeratedCount`, I'd use that instead. This can be found out if the `TryGetNonEnumeratedCount` method symbol is available on the enumerable type.
Probably use a type switch / pattern matching as fallback.

TimothyMakkison replied:
I'd thought of doing that but wasn't sure how to check which framework is being used. I could check if Enumerable has the corresponding method.

latonz commented on src/Riok.Mapperly/Descriptors/Mappings/ExistingTarget/EnsureCapacityBuilder.cs:
```diff
@@ -0,0 +1,131 @@
+{
+    public string? SourceAccessor { get; set; }
+    public string TargetAccessor { get; set; } = null!;
+    public TypeSyntax? CollectionType { get; set; }
```
I'd stick to the symbols here instead of using the syntax and just build the typesyntax in the build method, if it is required there.

latonz commented on test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:
```diff
@@ -565,6 +601,52 @@ public void EnumerableToCreatedQueue()
+            .Should()
+            .HaveMethodBody("MapToStack",
+                """
+                var target = new System.Collections.Generic.Stack<long>();
+                target.EnsureCapacity(source.Length + target.Count);
```
This is not yet optimal, the built in collections all have a ctor accepting a capacity parameter which would reduce the array allocations and could be used by Mapperly. However, this could also be adressed in a separate PR.

TimothyMakkison replied:
> However, this could also be adressed in a separate PR.

I'd planned on updating the issue after completing this PR. I was going to look at how you did this with dictionaries.

What concrete types is this needed for? Types like `Stack<T>`, `Queue<T>` have constructors that take `IEnumerable<T>`  so I was going to use that where possible.

latonz replied:
Sounds good to me. `Stack<T>`, `Queue<T>` are all, as I think `List<T>` and co are always mapped to via linq.

TimothyMakkison replied:
Should I make those changes a separate PR?

latonz replied:
I'd do it separately if possible.

latonz reviewed (changes requested):
Thank you for another great contribution 🎉 

I'd create a new namespace `Riok.Mapperly.Descriptors.Enumerables` for these two classes (`EnsureCapacityInfo` and `EnsureCapacityBuilder`)

TimothyMakkison commented:
- Added version specific `TryGetNonEnumeratedCount` - I've updated the integration test to test this.
- Only use symbols until ensure capacity is built.
- Refactored `EnsureCapacityBuilder` into 3 different versions - wasnt a fan of creating 5 files but it felt better than using ad-hoc discriminant union. I can convert it back to the previous version if you prefer.
- All generated code uses the global type.

latonz commented on src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilderIsType.cs:
```diff
@@ -0,0 +1,51 @@
+
+    public StatementSyntax BuildEnsureCapacityStatement(TypeMappingBuildContext ctx, ExpressionSyntax target)
+    {
+        var targetCount = MemberAccess(target, _targetAccessor);
```
To improve the readability add comments with how the generated code looks like.

latonz commented on src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityHelper.cs:
```diff
@@ -0,0 +1,83 @@
+        return false;
+    }
+
+    public static ExpressionStatementSyntax EnsureCapacityStatement(ExpressionSyntax target, ExpressionSyntax sourceCount, ExpressionSyntax targetCount)
```
I'd move this syntax conversion into an abstract base class EnsureCapacity to follow the concept of MappingBuilders and Mappings.

latonz commented on src/Riok.Mapperly/Descriptors/MappingBuilders/EnumerableMappingBuilder.cs:
```diff
@@ -74,18 +75,24 @@ public static class EnumerableMappingBuilder
 
         return null;
+
+        ForEachAddEnumerableExistingTargetMapping CreateForEach(string propertyName)
```
Is the parameter name `propertyName` correct here?

latonz commented on src/Riok.Mapperly/Helpers/SymbolExtensions.cs:
```diff
@@ -103,6 +103,29 @@ internal static bool HasImplicitInterfaceMethod(this ITypeSymbol symbol, INamedT
         return !methodInterImplementaton.ExplicitInterfaceImplementations.Any();
     }
 
+    internal static bool HasImplicitInterfaceProperty(this ITypeSymbol symbol, INamedTypeSymbol inter, string propertyName)
```
Do we need the limitation of properties here?
I'd probably prefer the approach of https://github.com/riok/mapperly/pull/341#discussion_r1167471503.

latonz commented on test/Riok.Mapperly.IntegrationTests/Dto/TestObjectDto.cs:
```diff
@@ -22,7 +22,7 @@ public TestObjectDto(int ctorValue, int unknownValue = 10, int ctorValue2 = 100)
 
 #if NET7_0_OR_GREATER
-        public required int RequiredValue { get; init; }
+        public int RequiredValue { get; init; }
```
Mhmmm, is this change expected?

latonz commented on test/Riok.Mapperly.IntegrationTests/Models/TestObject.cs:
```diff
@@ -21,7 +21,9 @@ public TestObject(int ctorValue, int unknownValue = 10, int ctorValue2 = 100)
 #if NET7_0_OR_GREATER
-        public required int RequiredValue { get; init; }
+
+        public int RequiredValue { get; init; }
```
Mhmmm, is this change expected?

TimothyMakkison replied:
mb I used Codemaid for linting and it still doesn't support modern C# 😞

latonz commented on test/Riok.Mapperly.IntegrationTests/Dto/TestObjectDto.cs:
```diff
@@ -53,6 +53,8 @@ public TestObjectDto(int ctorValue, int unknownValue = 10, int ctorValue2 = 100)
 
         public TestObjectNestedDto[]? NullableReadOnlyObjectCollection { get; set; }
 
+        public Stack<int> EnsureCapacityValue { get; set; } = new();
```
Is this separate test property really required? What additional test case does it test?

TimothyMakkison replied:
`Enumerable.TryGetNonEnumeratedCount` was added in .Net 6+. If a lower version is used a statement using pattern matching is created (`EnsureCapacityBuilderIsType`). I took advantage of the .Net Framework 4.8 integration test to ensure that this works.

I could probably remove `EnsureCapacityBuilderIsType` as its never used in .Net 6+

latonz replied:
Ah now I got it, this tests a mapping `IEnumerable<string>` => `Stack<int>` (which uses the type pattern matching / TryGetNonEnumeratedCount respectively), `Stack<string>` => `Stack<int>` tests the case for which the source count is available directly.
I'd keep `EnsureCapacityIsType` for now, as Mapperly tries to support .NET Framework as good as possible.

latonz reviewed (changes requested):
I like the new approach. 
I'd adjust the names to the general concept of Mapperly with MappingBuilders and Mappings:
> The idea is to use a MappingBuilder to construct a Mapping object. The Mapping then should only transform the given information into a syntax. The MappingBuilder should provide all the required information for the Mapping to do only the transformation.

I'd rename the `EnsureCapacityBuilder*` to `EnsureCapacity*` (as they are not builder in the sense of Mapperly). However, the `EnsureCapacityHelper` would be a builder.
`IEnusreCapacityBuilder` => `EnsureCapacity` (abstract base class)
`EnsureCapacityHelper` => `EnsureCapacityBuilder`
`EnsureCapacityBuilderNonEnumerated` => `EnsureCapacityNonEnumerated`
`EnsureCapacityBuilderSimple` => `EnsureCapacityMember`
`EnsureCapacityBuilderIsType` => `EnsureCapacityWellKnownTypes`

latonz commented on src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacity.cs:
```diff
@@ -0,0 +1,21 @@
+
+    public abstract StatementSyntax BuildEnsureCapacityStatement(TypeMappingBuildContext ctx, ExpressionSyntax target);
+
+    public static ExpressionStatementSyntax EnsureCapacityStatement(ExpressionSyntax target, ExpressionSyntax sourceCount, ExpressionSyntax targetCount)
```
This method should be protected, shouldn't it?

latonz commented on src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:
```diff
@@ -0,0 +1,81 @@
+            return true;
+        }
+
+        value.ImplementsGeneric(types.IReadOnlyCollectionT, CountMethodName, out var _, out var hasReadOnlyCount);
```
Nit: the `var` keyword in `out var _` is not needed, is it? (same on L64).

latonz commented on src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:
```diff
@@ -0,0 +1,81 @@
+public static class EnsureCapacityBuilder
+{
+    private const string EnsureCapacityName = "EnsureCapacity";
+    private const string CountMethodName = nameof(ICollection<object>.Count);
+    private const string LengthMethodName = nameof(Array.Length);
```
These are not really method but property names...

latonz commented on src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:
```diff
@@ -0,0 +1,81 @@
+    private const string LengthMethodName = nameof(Array.Length);
+    private const string TryGetNonEnumeratedCountMethodName = "TryGetNonEnumeratedCount";
+
+    public static EnsureCapacity? TryCreateEnsureCapacityBuilder(ITypeSymbol sourceType, ITypeSymbol targetType, WellKnownTypes types)
```
I'd simplify this method's name (it doesn't return a builder). Eg. `TryBuildEnsureCapacity` (would also align with `TryBuildMapping`).

latonz commented on src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacity.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    private const string EnsureCapacityName = "EnsureCapacity";
+
+    public abstract StatementSyntax BuildEnsureCapacityStatement(TypeMappingBuildContext ctx, ExpressionSyntax target);
```
I'd simplify this method name. There is only one build method in this class which is exposed and since the class is named `EnsureCapacity` it should be clear that it is an ensure capacity statement. IMO `Build` should be simple and understandable.

latonz commented on src/Riok.Mapperly/Helpers/SymbolExtensions.cs:
```diff
@@ -79,6 +79,49 @@ internal static IEnumerable<IPropertySymbol> GetAllAccessibleProperties(this ITy
+            return false;
+        }
+
+        var interfaceSymbol = genericIntf.GetMembers(symbolName).Single();
```
TBD: maybe switching to `First` instead of `Single` will be more resilient if a future release adds a method overload for a given method.

latonz commented on test/Riok.Mapperly.IntegrationTests/Models/TestObject.cs:
```diff
@@ -22,6 +22,7 @@ public TestObject(int ctorValue, int unknownValue = 10, int ctorValue2 = 100)
 
 #if NET7_0_OR_GREATER
         public required int RequiredValue { get; init; }
+
```
Nit: Probably revert this change?

latonz commented on test/Riok.Mapperly.Tests/Mapping/EnumerableTest.cs:
```diff
@@ -631,6 +631,47 @@ public void MapToExistingExplicitAddCustomCollection()
+        // should not create an EnsureCapacity statement when the target has an explicit count
+        var source = TestSourceBuilder.MapperWithBodyAndTypes(
+            "partial void Map(IEnumerable<int> source, B target);",
+            "class B : List<int> { int List<int>.Count { get; } }");
```
I don't get this test case, class B extends the List class, but implements the base class property explicit. This does not really work, does it?

TimothyMakkison replied:
Yeah this test is terrible for many reasons. I decided not to generate an EnsureCapacity statement when the target type has an `EnsureCapacity` method and an explicit length. This is because I deemed it to be an edge case and a complete pain to handle with `EnsureCapacityIsType`. 

For some reasons I made a (broken) test to ensure it won't generate an `EnsureCapacity` statement under these conditions.

I could either delete this test, fix it, or update the logic to insert a cast to cast the target when its `Count` is explicit.

latonz replied:
I'm okay with deleting this test.

latonz commented on test/Riok.Mapperly.IntegrationTests/_snapshots/NET_48/StaticMapperTest.SnapshotGeneratedSource.verified.cs:
```diff
@@ -319,7 +323,18 @@ public static partial class StaticTestMapper
+        private static global::System.Collections.Generic.Stack<int> MapToStack(global::System.Collections.Generic.IEnumerable<string> source)
+        {
+            var target = new global::System.Collections.Generic.Stack<int>();
+            foreach (var item in source)
```
Shouldn't this be the point where the type tests should be emitted? As mentioned in another comment, I'm also ok with removing the `EnsureCapacityIsType` but then this integration test property could also be removed.

latonz commented on src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:
```diff
@@ -0,0 +1,81 @@
+        if (TryGetNonEnumeratedCount(sourceType, types, out var sourceSizeProperty))
+            return new EnsureCapacityMember(targetSizeProperty, sourceSizeProperty);
+
+        sourceType.ImplementsGeneric(types.IEnumerableT, out var iEnumerable);
```
Instead of asserting `iEnumerable` is non-null later, I'd if + return null here:
```c#
if (!sourceType.ImplementsGeneric(types.IEnumerableT, out var iEnumerable))
    return null;
```

latonz reviewed (changes requested):
Looks pretty good 😊 Thank you a lot for your efforts!

A few nitpicks left and then this should be ready to merge 🚀

latonz commented on src/Riok.Mapperly/Helpers/SymbolExtensions.cs:
```diff
@@ -79,6 +79,49 @@ internal static IEnumerable<IPropertySymbol> GetAllAccessibleProperties(this ITy
+        return true;
+    }
+
     internal static bool HasImplicitInterfaceMethod(this ITypeSymbol symbol, INamedTypeSymbol inter, string methodName)
```
This could now be simplified a lot, couldn't it? (eg. `symbol.ImplementsGeneric(inter, methodName, out var isExplicit) && !isExplicit`

TimothyMakkison commented:
Turns out `EnsureCapacity` wasn't added until [.Net 6+](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.list-1.ensurecapacity?view=net-8.0#system-collections-generic-list-1-ensurecapacity(system-int32)), so `EnsureCapacityIsType` and the integration test are not needed.

latonz commented on src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:
```diff
@@ -0,0 +1,69 @@
+        if (TryGetNonEnumeratedCount(sourceType, types, out var sourceSizeProperty))
+            return new EnsureCapacityMember(targetSizeProperty, sourceSizeProperty);
+
+        if (!sourceType.ImplementsGeneric(types.IEnumerableT, out var iEnumerable))
+            return null;
```
Is this ever reached? `TryBuildEnsureCapacity` is only called for enumerables, isn't it?

TimothyMakkison replied:
I used ImplementsGeneric to get the typed version of `IEnumerableT`. I originally didn't use this redundant null, but changed it when you [suggested I remove the null suppression](https://github.com/riok/mapperly/pull/312#discussion_r1167799422)

latonz commented on src/Riok.Mapperly/Descriptors/Enumerables/EnsureCapacity/EnsureCapacityBuilder.cs:
```diff
@@ -0,0 +1,69 @@
+                                                        .FirstOrDefault(x => x.ReturnType.SpecialType == SpecialType.System_Boolean && x.IsStatic && x.Parameters.Length == 2 && x.IsGenericMethod);
+
+        // if source does not have a count use GetNonEnumeratedCount, calling EnusureCapacity if count is available
+        var typedNonEnumeratedCount = nonEnumeratedCountMethod.Construct(iEnumerable.TypeArguments.ToArray());
```
This leads to a null ref exception on .NET Framework... add a null guard:
```c#
if (nonEnumeratedCountMethod == null)
    return null;

...
```

TimothyMakkison replied:
.NET Framework can't get to this section, [line 23](https://github.com/riok/mapperly/pull/312/files/f19efb7c50ed837ab5986cf93b4ac58f64072d96#diff-3f8932b41774a45b65c962c4ae183d71993fe8112d68981d3ab20f2a1d53172fR23) prevents non .NET 6+ code from getting here because they don't have `EnsureCapacity` available yet.
